README.md
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Binary Defense Systems Auto-Enroll for OSSEC
Written by: David Kennedy - BDS
Version 1.5.1
Supported Systems: Linux, Windows
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The auto ossec enrollment will automatically provision OSSEC agents for both Linux and Windows. This is accomplished through a customized protocol
that interfaces with the ossec server and the automatic pairing of the server. 


Descriptions:

auto_server.py - this file contains the server to house the protocol - the port needed for this is 9654. The agents will communicate with the agents to this server script that is running. This server script should be placed in an automatic init script and through watchdog in order to ensure its always running. This script needs to be placed on the OSSEC server itself.

auto_ossec.exe and auto_ossec.py - auto_ossec.exe is to run on Windows, and auto_ossec.py to run on Linux. When running the tool you will need to issue auto_ossec.exe <ip_address_of_auto_server> - this will be the IP address of the server itself.

auto_ossec.bin - precompiled for Linux with all libraries bundies in (including python-crypto)

Deployment Instructions:

Install OSSEC server on a system. Ensure that auto_server.py is automatically started when reboot occurs, and watchdog in order to ensure its always running. Also ensure no iptables conflict with port 9654 - this port is needed for the two way communication.

Install OSSEC on a Linux or Windows system as an agent. Then run auto_ossec.exe, auto_ossec.bin, or auto_ossec.py with the IP address of the SERVER that is running auto_server.py. This will automatically pair the instances of OSSEC.

NOTE THAT ALL OF THESE NEED TO BE PERFORMED WITH ROOT OR ADMINISTRATIVE LEVEL PERMISSIONS. THIS WILL FAIL IF IT IS NOT INSTALLED WITH ADMIN PRIVS.

Mass Deployment Instructions:

Create a deployment package that first installs the OSSEC binary or tar ball from (http://www.ossec.net/?page_id=19). Once the install completes, run the auto_ossec <server_ip> and you are finished. Services will automatically restart.

Ports Needed: 9654

What the server should look like when you run it in an interactive interface:

Client connected with  ('192.168.170.165', 50662)
[*] Provisioned new key for hostname: STRONGHOLD-WIN8 with IP of: 192.168.170.165
[*] Sending new key to 192.168.170.165: 8zlUouJ7yVOvt06Er8yx1zTchy5VQklfovu4SXW3GX7X8gH5tPIZ1104wvleQoZmJ9Hod++ByQtgNSLrQV7Z7rsRZLhCS9hFxPwRTZu6JC80EUXJ4yuTqFPHf9L2QuDjelP0yUvFFExf0xm7czlmDVH6/VKRdms1nL8+mwC9S81aZ0IOGpZuIMbIwiyeVxyBpctCk0Qd5CHoVZaKpAWTtA==
Pairing complete. Terminating connection to client.

## Linux Automatic Installation

You can now automatically install Linux as part of the auto_ossec.bin and auto_ossec.py installer. If the proper variable is specified, auto_ossec will automatically download OSSEC from the Internet, install it for you as an agent, and configure the server and keypairs for you.

In order to automatically install Linux, run the following command as root/sudo:

./auto_ossec.bin <server_ip> url=urlto/ossec.tar.gz

The server IP address is the IP address of your OSSEC server installation. The url=<site> specifies that you want to automatically install. You can also specify a wildcard for your hostname of the AGENT. To do this you can type:

./auto_ossec.bin <server_ip> * url=https://bintray.com/artifact/download/ossec/ossec-hids/ossec-hids-2.8.3.tar.gz

This will automatically install OSSEC through the Internet and specify a wildcard for the IP address of the agent. This is useful when installing agents on dynamic IP addresses.

## Regular Linux Installation

If you install OSSEC regularly on Linux, you can just install OSSEC normally on Linux. Be sure to specify the right server IP address of where your OSSEC server is at and the IP address of where auto_server.py is running (the server enroller).

./auto_ossec.bin <server_ip>

## Install on Windows

For Windows, install OSSEC normally - since it is an MSI you should install this silently. Once OSSEC is installed, run:

auto_ossec.exe <server_ip>

This will automatically update your OSSEC config file with the server IP address and do the magic needed to pair them. You can also use a * (wildcard):

auto_ossec.exe <server_ip> *

This will be useful if your system changes IP addresses frequently (dynamic DNS)

### Supported Operating Systems

Linux
Windows




CHANGELOG.txt
~~~~~~~~~~~~~~~~
version 1.4
~~~~~~~~~~~~~~~~

* added allow reuse for threadedtcp server to allow binding immediately when the port is shut down
* added linux binary to allow native deployment of linux without python-crypto needed
* formatted in autopep8 for consistency on tab indentation
* added ability to use star as an IP address


auto_server.py
#!/usr/bin/python
#
# This is the ossec auto enrollment server daemon. This should be put under supervisor to ensure health and stability.
#
#
# Works with Alienvault and Standlone OSSEC installs
#
# Will listen on port 9654 for an incoming challege
#
import socketserver
from threading import Thread
import subprocess
import sys

# check python crypto library
try:
    from Crypto.Cipher import AES

except ImportError:
    print(
        "[!] ERROR: python-crypto not installed. Run 'apt-get install python-pycrypto pexpect' to fix.")
    sys.exit()

import base64
import _thread

# check pexpect library
try:
    import pexpect
except ImportError:
    print("[!] ERROR: pexpect not installed. Run apt-get install pexpect to fix.")
    sys.exit()

import time
import socket
import os


class service(socketserver.BaseRequestHandler):

    def handle(self):
        # parse OSSEC hids client certificate
        def parse_client(hostname, ipaddr):
            child = pexpect.spawn("/var/ossec/bin/manage_agents")
            child.expect("Choose your action")
            child.sendline("a")
            child.expect("for the new agent")
            child.sendline(hostname)
            i = child.expect(
                ['IP Address of the new agent', 'already present'])

            # if we haven't already added the hostname
            if i == 0:
                child.sendline(ipaddr)
                child.expect("for the new agent")
                child.sendline("")
                for line in child:
                    # pull id
                    if "[" in line:
                        id = line.replace(
                            "[", "").replace("]", "").replace(":", "").rstrip()
                    break
                child.expect("Confirm adding it?")
                child.sendline("y")
                child.sendline("")
                child.sendline("q")
                child.close()

                child = pexpect.spawn(
                    "/var/ossec/bin/manage_agents -e %s" % (id))
                for line in child:
                    key = line.rstrip()

                return key

            # if we have a duplicate hostname
            else:
                child.close()
                child = pexpect.spawn("/var/ossec/bin/manage_agents -l")
                for line in child:
                    line = line.rstrip()
                    if hostname in line:
                        id = line.split(",")[0].replace(
                            "ID: ", "").replace("   ", "").rstrip()
                        break
                child.close()
                subprocess.Popen("/var/ossec/bin/manage_agents -r %s" %
                                 (id), stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True).wait()
                return 0

        # main AES encrypt and decrypt function with 32 block size padding
        def aescall(secret, data, format):

            # padding and block size
            PADDING = '{'
            BLOCK_SIZE = 32

            # one-liner to sufficiently pad the text to be encrypted
            pad = lambda s: s + (BLOCK_SIZE - len(s) % BLOCK_SIZE) * PADDING

            # random value here to randomize builds
            a = 50 * 5

            # one-liners to encrypt/encode and decrypt/decode a string
            # encrypt with AES, encode with base64
            EncodeAES = lambda c, s: base64.b64encode(c.encrypt(pad(s)))
            DecodeAES = lambda c, e: c.decrypt(
                base64.b64decode(e)).rstrip(PADDING)

            cipher = AES.new(secret)

            if format == "encrypt":
                aes = EncodeAES(cipher, data)
                return str(aes)

            if format == "decrypt":
                aes = DecodeAES(cipher, data)
                return str(aes)

        # recommend changing this - if you do, change auto_ossec.py as well - -
        # would recommend this is the default published to git
        secret = "(3j+-sa!333hNA2u3h@*!~h~2&^lk<!B"
        print("Client connected with ", self.client_address)
        try:
            data = self.request.recv(1024)
            if data != "":
                try:
                    data = aescall(secret, data, "decrypt")

                    # if this section clears -we know that it is a legit
                    # request, has been decrypted and we're ready to rock
                    if "BDSOSSEC" in data:

                        # if we are using star IP addresses
                        if "BDSOSSEC*" in data:
                            star = 1
                        else:
                            star = 0

                        # write a lock file to check later on with our threaded
                        # process to restart OSSEC if needed every 10 minutes -
                        # if lock file is present then it will trigger a
                        # restart of OSSEC server
                        if not os.path.isfile("lock"):
                            filewrite = file("lock", "w")
                            filewrite.write("lock")
                            filewrite.close()

                        # strip identifier
                        data = data.replace(
                            "BDSOSSEC*", "").replace("BDSOSSEC", "")
                        hostname = data

                        # pull the true IP, not the NATed one if they are using
                        # VMWare
                        if star == 0:
                            ipaddr = self.client_address[0]
                        else:
                            ipaddr = "*"

                        # here if the hostname was already used, we need to
                        # remove it and call it again
                        data = parse_client(hostname, ipaddr)
                        if data == 0:
                            data = parse_client(hostname, ipaddr)
                        print("[*] Provisioned new key for hostname: %s with IP of: %s" %
                              (hostname, ipaddr))
                        data = aescall(secret, data, "encrypt")
                        print("[*] Sending new key to %s: " % (ipaddr) + data)
                        self.request.send(data)

                except Exception as e:
                    print(e)
                    pass

        except Exception as e:
            print(e)
            pass

        print("Pairing complete. Terminating connection to client.")
        self.request.close()

# this waits 5 minutes to check if new ossec agents have been deployed, if
# so it restarts the server


def ossec_monitor():
    while 1:
        time.sleep(300)
        if os.path.isfile("lock"):
            os.remove("lock")
            print(
                "[*] New OSSEC agent added - triggering restart of service to add..")
            subprocess.Popen("service ossec restart", stdout=subprocess.PIPE,
                            stderr=subprocess.PIPE, shell=True).wait()


class ThreadedTCPServer(socketserver.ThreadingMixIn, socketserver.TCPServer):
    pass

print("[*] The auto enrollment OSSEC Server is now listening on 9654")
s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
# set is so that when we cancel out we can reuse port
s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
# bind to all interfaces on port 10900

ThreadedTCPServer.allow_reuse_address = True
t = ThreadedTCPServer(('', 9654), service)
# start the server and listen forever
try:
    # start a threaded counter
    _thread.start_new_thread(ossec_monitor, ())

    t.serve_forever()

except KeyboardInterrupt:
    print("[*] Exiting the automatic enrollment OSSEC daemon")


auto_ossec.py
#!/usr/bin/python
#
# This is the initial release of the ossec_client.py which can be byte compiled into an exe. This will
# connect to the ossec_auto.py daemon that will automatically issue a key in order to pair the OSSEC
# HIDS.
#
# Also works with AlienVault.
#
# NOTE that you NEED to change the host = '' to the appropriate IP address of the OSSEC server and
# where the ossec_auto.py daemon is running.
#

import platform
import base64
import socket
import sys
import os
import subprocess
import time
try: import urllib.request as urllib
except: import urllib

# try to import python-crypto
try:
    from Crypto.Cipher import AES

except ImportError:
    print (
        "[!] You need python-crypto in order for this module to work. If this is Ubuntu/Redhat - package name is python-crypto")
    sys.exit()

# check platform specific installs
installer = ""
if platform.system() == "Linux":
    installer = "Linux"

if platform.system() == "Windows":
    installer = "Windows"

if installer == "":
    print (
        "[!] Unable to determine operating system. Only supports Linux and Windows. Exiting..")
    sys.exit()


#
# NEED TO DEFINE THIS AS THE OSSEC SERVER HOST THAT IS RUNNING SERVER.PY
#
star = ""
autoinstall = ""
version_name = ""
try:
    host = sys.argv[1]
    try: 
        test = sys.argv[2]
        if "*" in sys.argv[2]: star = sys.argv[2]
        else: autoinstall = sys.argv[2]

    except: pass

    try: 
        test = sys.argv[3]
        if "*" in sys.argv[3]: star = sys.argv[3]
        else: autoinstall = sys.argv[3]

    except: pass

except IndexError:
    print ("""
*********************************************************************************************
Binary Defense Systems OSSEC Auto Enrollment

In order for this to work, you need to point
auto_ossec.exe to the OSSEC server that is
listening. Note that default port is 9654
but this can be changed in the source.

Note that if you specify optional *, this will
place a * for the IP address in the config and
allow any IP address (for dynamic IP addresses).

Also note if you specify url=<site> at the end, this is for
Linux only, it will automatically download and install
OSSEC for you and configure it based on the server-ip.
You do not need to do a * before

Example: auto_ossec.exe/.bin 192.168.5.5 * url=https://bintray.com/etc/etc/ossec-hids.tar.gz
Example2: auto_ossec.bin 192.168.5.5 url=https://somewebsite.com/ossec-hids-2.8.3.tar.gz
Usage: auto_ossec.exe <server_ip> <optional: *> <optional: url>

Example URL: https://bintray.com/artifact/download/ossec/ossec-hids/ossec-hids-2.8.3.tar.gz

********************************************************************************************
		""")
    sys.exit()

# url for OSSEC HERE
if "url=" in autoinstall: 
    url = autoinstall.replace("url=", "").replace('"', "", 2)
    version_name = url.split("/ossec-hids/")[1].replace(".tar.gz", "")

# download ossec
def _download_ossec(url):
    ossec_file = urllib.urlopen(url).read()
    filewrite = open("/tmp/ossec.tar.gz", "wb")
    filewrite.write(ossec_file)
    filewrite.close()

# install ossec once downloaded
def _installossec(serverip,version_name):
    cwd = os.getcwd()
    os.chdir("/tmp/")
    subprocess.Popen("tar -zxvf ossec.tar.gz;rm ossec.tar.gz", shell=True).wait()

    #####
    #####
    ##### CHANGE THESE IF YOU WANT DIFFERENT CONFIG OPTIONS - read: http://ossec-docs.readthedocs.io/en/latest/manual/installation/install-source-unattended.html
    #####
    #####
    ossec_preload = ('''
                    USER_LANGUAGE="en"
                    USER_NO_STOP="y"
                    USER_INSTALL_TYPE="agent"
                    USER_DIR="/var/ossec"
                    USER_ENABLE_ACTIVE_RESPONSE="n"
                    USER_ENABLE_SYSCHECK="y"
                    USER_ENABLE_ROOTCHECK="y"
                    USER_UPDATE_RULES="y"
                    USER_AGENT_SERVER_IP="%s"
                    USER_ENABLE_EMAIL="n"
                    USER_ENABLE_FIREWALL_RESPONSE="n"
                    ''' % (serverip))

    filewrite = open("/tmp/%s/etc/preloaded-vars.conf" % (version_name), "w")
    filewrite.write(ossec_preload)
    filewrite.close()
    subprocess.Popen("cd %s;chmod +x;./install.sh" % (version_name), shell=True).wait()
    subprocess.Popen("rm -rf /tmp/%s" % (version_name), shell=True).wait()

# this is the auto installation process here
if installer == "Linux":
    if "url=" in autoinstall:
        print("[*] Automatically installing OSSEC on Linux for you with version: " + (version_name))
        _download_ossec(url)
        _installossec(host,version_name)

def aescall(secret, data, format):

    # padding and block size
    PADDING = '{'
    BLOCK_SIZE = 32
    # one-liner to sufficiently pad the text to be encrypted
    pad = lambda s: s + (BLOCK_SIZE - len(s) % BLOCK_SIZE) * PADDING

    # random value here to randomize builds
    a = 50 * 5

    # one-liners to encrypt/encode and decrypt/decode a string
    EncodeAES = lambda c, s: base64.b64encode(c.encrypt(pad(s)))
    DecodeAES = lambda c, e: c.decrypt(base64.b64decode(e)).rstrip(PADDING)

    cipher = AES.new(secret)

    if format == "encrypt":
        aes = EncodeAES(cipher, data)
        return str(aes)

    if format == "decrypt":
        aes = DecodeAES(cipher, data)
        return str(aes)

# this will grab the hostname and ip address and return it


def grab_info():
    try:
        hostname = socket.gethostname()
        return hostname  # + " " + ipaddr

    except Exception:
        sys.exit()
try:
    # secret key - if you change this you must change on ossec_auto server -
    # would recommend this is the default published to git
    secret = "(3j+-sa!333hNA2u3h@*!~h~2&^lk<!B"
    # port for daemon
    port = 9654
    # general length size of socket
    size = 1024

    # loop through in case server isnt reachable
    while 1:
        try:
            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            s.connect((host, port))
            break

        except Exception:
            print (
                "[!] Unable to connect to destination server. Re-trying in 10 seconds.")
            time.sleep(10)
            pass

    print(("[*] Connected to auto enrollment server at IP: " + host))
    # grab host info needed for ossec
    data = grab_info()

    # encrypt the data
    if star == "*":
        data = "BDSOSSEC*" + data.rstrip()
    else:
        data = "BDSOSSEC" + data.rstrip()
    data = aescall(secret, data, "encrypt")
    print (
        "[*] Pulled hostname and IP, encrypted data, and now sending to server.")
    s.send(data)
    data = s.recv(size)
    # this is our ossec key
    print (
        "[*] We received our new pairing key for OSSEC, closing server connection.")
    data = aescall(secret, data, "decrypt")
    # close socket
    s.close()

    # path variables for OSSEC
    if os.path.isdir("C:\\Program Files (x86)\\ossec-agent"):
        path = "C:\\Program Files (x86)\\ossec-agent"
    if os.path.isdir("C:\\Program Files\\ossec-agent"):
        path = "C:\\Program Files\\ossec-agent"
    if os.path.isdir("/var/ossec/"):
        path = "/var/ossec/"
    if path == "":
        sys.exit()
    print ("[*] Removing any old keys.")
    os.chdir(path)

    if installer == "Windows":
        if os.path.isfile("client.keys"):
            os.remove("client.keys")
        # import the key with the key presented from the server daemon
        filewrite = open(path + "\\client.keys", "w")

    if installer == "Linux":
        if os.path.isfile(path + "/etc/client.keys"):
            os.remove("etc/client.keys")
        filewrite = open(path + "/etc/client.keys", "w")
    data = base64.b64decode(data)
    filewrite.write(data)
    filewrite.close()
    print ("[*] Successfully imported the new pairing key.")
    print ("[*] Stopping the OSSEC service, just in case its running.")
    # stop the service if it is
    if installer == "Windows":
        subprocess.Popen('net stop "OSSEC HIDS"', stdout=subprocess.PIPE,
                         stderr=subprocess.PIPE, shell=True).wait()

    if installer == "Linux":
        subprocess.Popen("service ossec stop", stdout=subprocess.PIPE,
                         stderr=subprocess.PIPE, shell=True).wait()

    # make sure we modify the ossec.conf
    if installer == "Windows":
        print ("[*] Modifying ossec.conf to incorporate server host IP address.")
        data = open(path + "\\ossec.conf", "r").read()
        if not "<server-ip>%s</server-ip>" % (host) in data:
            filewrite = open(path + "\\ossec.conf", "a")
            filewrite.write("\n")
            filewrite.write(" <ossec_config>")
            filewrite.write("   <client>")
            filewrite.write("      <server-ip>%s</server-ip>" % (host))
            filewrite.write("   </client>")
            filewrite.write(" </ossec_config>")
            filewrite.close()

    # start the service
    if installer == "Windows":
        subprocess.Popen('net start "OSSEC HIDS"', stdout=subprocess.PIPE,
                         stderr=subprocess.PIPE, shell=True).wait()

    if installer == "Linux":
        subprocess.Popen("service ossec start", stdout=subprocess.PIPE,
                         stderr=subprocess.PIPE, shell=True).wait()

    print (
        "[*] Finished. Started the OSSEC service. Auto Enrollment for OSSEC is now finished.")


except KeyboardInterrupt:
    print ("Sounds good.. Abording Auto-OSSEC...")
    sys.exit()
except Exception as error:
    print (
        "[*] Something did not complete. Does this system have Internet access?")
    print (error)
